File: pset6/credit/credit.py
from cs50 import get_string


#First Check if card is a valid number
#Check what bank the card belongs to

#Functions must be defined before using


#Checks if number is a valid credit card number
def LuhnsAlg(cardN):
    sum1, sum2 = 0, 0
    
    #allows to alternate between sum 1 and sum 2 starting from end of list
    posCheck = True
    
    
    #iterates through entire credit number and posCheck allows to get every other number
    for i in range(len(cardN), 0, -1):
        
        if posCheck == True:
            sum2 += int(cardN[i - 1])
            posCheck = False
        else:
            temp = 2 * int(cardN[i - 1])
            if temp // 10 == 0:
                sum1 += temp
            else:
                sum1 += (temp % 10) + (temp // 10)
            posCheck = True
    
    if (sum1 + sum2) % 10 == 0:
        return True
    else: 
        return False
        
        
    
#Checks if what bank and if card is a valid bank card  
def bankCheck(cardN):
    
    if len(cardN) == 15 and cardN[0:2] in ['34', '37']:
        return "AMEX"
    elif len(cardN) == 16 and cardN[0:2] in ['51', '52', '53', '54', '55']:
        return "MASTERCARD"
    elif (len(cardN) in [13, 16]) and cardN[0] == '4':
        return "VISA"
    else:
        return "INVALID"
        
        
        
credit = get_string("Number: ")
if LuhnsAlg(credit):
    print(bankCheck(credit))
else:
    print("INVALID")

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long credit = get_long("Number: ");
    long creditC = credit;

    int check = 1;
    int checkSumEven = 0;
    int checkSumOdd = 0;
    
    // Iterates through credit number following Luhn's Alg
    do
    {
        if (check % 2 == 0)
        {
            int digits = 2 * (credit % 10);
           
            do
            {
               checkSumEven = checkSumEven + (digits % 10);
               digits = digits / 10;
            }while(digits != 0);
        }
        
        else
        {
            checkSumOdd = checkSumOdd + (credit % 10);
            // printf("Odd: %i\n",checkSumOdd);
            // printf("credit number: %li\n",credit);
        }
        credit = credit / 10;
        check++;
    }while(credit != 0);
    
    int total = checkSumEven + checkSumOdd;
    
    // Checks if credit card is valid and assigns correct credit company for each valid credit card number
    if (total % 10 == 0)
    {
        if (creditC / 10000000000000 == 34 || creditC / 10000000000000 == 37)
        {
            printf("AMEX\n");
        }
        else if (creditC / 1000000000000 == 4 || creditC / 1000000000000000 == 4)
        {
            printf("VISA\n");
        }
        else if (creditC / 100000000000000 == 51 || creditC / 100000000000000 == 52 || creditC / 100000000000000 == 53 
        || creditC / 100000000000000 == 54 || creditC / 100000000000000 == 55)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}